File: Android_Firebase_Demo/app/src/main/java/firebase_demo/fg/com/android_firebase_demo/MainActivity.java
package firebase_demo.fg.com.android_firebase_demo;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.widget.ListView;

import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

public class MainActivity extends AppCompatActivity {

    private DatabaseReference mDatabase;
    private DataListAdapter listAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        this.mDatabase = FirebaseDatabase.getInstance().getReference();

        ListView list = (ListView) this.findViewById(R.id.dataList);
        this.listAdapter = new DataListAdapter(this, R.layout.list_view_cell);
        list.setAdapter(listAdapter);
    }

    @Override
    protected void onStart() {
        super.onStart();

        this.getSingleDataShot();

    }

    protected void getSingleDataShot() {
        ChildEventListener childEventListener = new ChildEventListener() {
            @Override
            public void onChildAdded(DataSnapshot dataSnapshot, String previousChildName) {
                Log.d("ADDED", "onChildAdded:" + dataSnapshot.getKey());
                User newUser = new User((String) dataSnapshot.getValue(), (String) dataSnapshot.getKey());
                if (!(listAdapter.getUserList().contains(newUser))) {
                    listAdapter.addUser(newUser);
                    listAdapter.notifyDataSetChanged();
                }
            }

            @Override
            public void onChildMoved(DataSnapshot dataSnapshot, String s) {
                Log.d("REMOVED", "");
            }

            @Override
            public void onChildRemoved(DataSnapshot dataSnapshot) {
                Log.d("DELETE : ", " ");
                listAdapter.removeUser(dataSnapshot.getKey());
                listAdapter.notifyDataSetChanged();
            }

            @Override
            public void onChildChanged(DataSnapshot dataSnapshot, String s) {
                Log.d("CHANGED", "");
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        };


        mDatabase.addChildEventListener(childEventListener);
    }
}

File: Android_Firebase_Demo/app/src/main/java/com/example/floriangbh/firebase/MainActivity.kt
package com.example.floriangbh.firebase

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import com.google.firebase.database.DatabaseReference

import kotlinx.android.synthetic.main.activity_main.*
import com.google.firebase.database.FirebaseDatabase



class MainActivity : AppCompatActivity() {

    var itemAdapter: ItemAdapter? = null
    var items: ArrayList<ItemData> = ArrayList()
    var databaseReference: DatabaseReference? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val database = FirebaseDatabase.getInstance()
        databaseReference = database.reference

        item_recycler.layoutManager = LinearLayoutManager(this)
        itemAdapter = ItemAdapter(items, { partItem : ItemData -> itemClicked(partItem) })
        itemAdapter.let { item_recycler.adapter = it }
    }

    override fun onStart() {
        super.onStart()

        val adapter = itemAdapter ?: return
        var databaseRef = databaseReference ?: return
        databaseRef.addChildEventListener(FirebaseEventListener(adapter))
    }

    fun itemClicked(item: ItemData) {
        println(item.id)
    }
}
